backend/src/apis/snapshot_data.rs

```diff
@@ -0,0 +1,249 @@
+{
+    pub fn new(
+        exchange_id: &str,
+        entry_csv: &str,
+        asset_csv: &str,
+    ) -> Result<SnapshotData<LEVELS, L, N_ASSETS, N_BYTES, K>, Box<dyn std::error::Error>> {
+        let assets = parse_csv_to_assets(asset_csv).unwrap();
+        let mst: MerkleSumTree<N_ASSETS> = MerkleSumTree::<N_ASSETS>::new(entry_csv).unwrap();
+
+        let user_proofs = HashMap::<u64, InclusionProof>::new();
+
+        Ok(SnapshotData {
+            exchange_id: exchange_id.to_owned(),
+            mst,
+            assets,
+            proofs_of_inclusion: user_proofs,
+            proof_of_solvency: None,
+        })
+    }
+
+    fn generate_inclusion_proof(&self, user_index: usize) -> Result<InclusionProof, &'static str> {
+        // TODO: fetch pk from outside. For now, we generate them here
+        let circuit = MstInclusionCircuit::<LEVELS, L, N_ASSETS>::init_empty();
+        let params = generate_setup_params(K);
+
+        let vk = keygen_vk(&params, &circuit).expect("vk generation should not fail");
+        let pk = keygen_pk(&params, vk.clone(), &circuit).expect("pk generation should not fail");
+
+        // Only now we can instantiate the circuit with the actual inputs
+        // let circuit = MstInclusionCircuit::<LEVELS, L, N_ASSETS>::init(entry_csv, user_index);
+        let proof = self.mst.generate_proof(user_index).unwrap();
+
+        let circuit = MstInclusionCircuit::<LEVELS, L, N_ASSETS> {
+            leaf_hash: proof.entry.compute_leaf().hash,
+            leaf_balances: proof
+                .entry
+                .balances()
+                .iter()
+                .map(big_int_to_fp)
+                .collect::<Vec<_>>(),
+            path_element_hashes: proof.sibling_hashes,
+            path_element_balances: proof.sibling_sums,
+            path_indices: proof.path_indices,
+            root_hash: proof.root_hash,
+        };
+
+        let instances = circuit.instances().clone();
+        let proof = full_prover(&params, &pk, circuit.clone(), instances.clone());
+
+        Ok(InclusionProof {
+            leaf_hash: instances[0][0],
+            vk: vk.to_bytes(halo2_proofs::SerdeFormat::RawBytes),
+            proof,
+        })
+    }
+
+    pub fn generate_solvency_proof(&mut self) -> Result<(), &'static str> {
+        // Prepare public inputs for solvency
+        let mut assets_sum = [Fp::from(0u64); N_ASSETS];
+        let asset_names = self
+            .assets
+            .iter()
+            .map(|asset| asset.name.clone())
+            .collect::<Vec<String>>();
+
+        for asset in &self.assets {
+            let index = asset_names.iter().position(|x| *x == asset.name).unwrap();
+            assets_sum[index] = asset.sum_balances;
+        }
+
+        // generate solvency proof
+        // TODO: fetch pk from outside. For now, we generate them here
+        let circuit = SolvencyCircuit::<L, N_ASSETS, N_BYTES>::init_empty();
+        let params = generate_setup_params(K);
```

Comment:
Same as before, I think that `pk` of Solvency Circuit can be generated at `init()` and stored somewhere in the `Snapshot`

Reply:
Does the `pk` is needed for the on-chain verifier?

Reply:
Yes, it is needed!

Reply:
I believe it's acceptable to generate the solvency proof upon receiving a request through the 'generate_solvency_proof' method. 
In any case, I have updated the `SolvenctyProof` in [309c563](https://github.com/summa-dev/summa-solvency/pull/88/commits/309c563d3c7862aee279b07ee035095c198c3792)

Reply:
I still see an issue here. 

The smart contract verifier should be already deployed when the CEX calls `generate_solvency_proof`. In particular, the smart contract requires the `vk` of the Circuit in order to be deployed. Therefore, it will only accept proof generated from the corresponding `pk`. 

In the current implementation `vk` and `pk` are generated inside `generate_solvency_proof` so the proof won't get verified by the smart contract because it is generated using a wrong proving key. 

I suggest here an implementation that I hope would make our life easier:

- Replace `MstParamsAndKeys` struct but a more abstract struct `CircuitArtifacts` that stores `pk` and, optionally, `vk` of a Circuit. Since it can store any proving and verification key, it no longer needs any generic parameters such as LEVELS, L ...
- Modify `new` to fetch `pk` for `MstInclusionCircuit` from storage, create a new instance of `CircuitArtifacts` and add it to Snapshot data. The `vk` can be retrieved from `pk` and added to the instance too. Then do the same for `SolvencyCircuit`, here we don't need the `vk`
- Since we don't have to call `generate_setup_params` any more, we can remove at all the need for `K` and any issue related to that. 

Note that in the latest commit to #80, I added the `pk` for the circuits to the `backend/artifacts` folder. In particular, you can see that the file name is based on the value assigned to the generics of the circuit. For example `mst_inclusion_15_6_2.pk` is the proving key for a `MstInclusionCircuit` of LEVELS=15, L=6, N_ASSETS=2

Reply:
There's still a problem here. If we generate a new random params here, the prover will fail, since the `pk` that we are fetching from `pk_path` was generated using some different params. We should make sure that the `params` used to deploy the verifier contract are the same as the ones used to extract the `pk` used to generate the proof

Reply:
Updated at 71d4b19e5a87d3d236622d32bb9831840767be00